File: mrbgems/mruby-plato-gpio-grpeach/test/gpio.rb
assert('GPIO - class') do
  assert_kind_of(Class, Plato::GPIO)
  assert_equal(Object, Plato::GPIO.superclass)
end

assert('GPIO::D0-D15') do
  assert_equal(0, Plato::GPIO::D0)
  assert_equal(1, Plato::GPIO::D1)
  assert_equal(2, Plato::GPIO::D2)
  assert_equal(3, Plato::GPIO::D3)
  assert_equal(4, Plato::GPIO::D4)
  assert_equal(5, Plato::GPIO::D5)
  assert_equal(6, Plato::GPIO::D6)
  assert_equal(7, Plato::GPIO::D7)
  assert_equal(8, Plato::GPIO::D8)
  assert_equal(9, Plato::GPIO::D9)
  assert_equal(10, Plato::GPIO::D10)
  assert_equal(11, Plato::GPIO::D11)
  assert_equal(12, Plato::GPIO::D12)
  assert_equal(13, Plato::GPIO::D13)
  assert_equal(14, Plato::GPIO::D14)
  assert_equal(15, Plato::GPIO::D15)
end

assert('GPIO::A0-A5') do
  assert_equal(16, Plato::GPIO::A0)
  assert_equal(17, Plato::GPIO::A1)
  assert_equal(18, Plato::GPIO::A2)
  assert_equal(19, Plato::GPIO::A3)
  assert_equal(20, Plato::GPIO::A4)
  assert_equal(21, Plato::GPIO::A5)
end

assert('GPIO::LED1-LED4') do
  assert_equal(22, Plato::GPIO::LED1)
  assert_equal(23, Plato::GPIO::LED2)
  assert_equal(24, Plato::GPIO::LED3)
  assert_equal(25, Plato::GPIO::LED4)
end

assert('GPIO::BUTTON1') do
  assert_equal(26, Plato::GPIO::BUTTON1)
end

File: mrbgems/mruby-plato-gpio-grpeach/src/gpio.c
#include "mruby.h"
#include "mruby/variable.h"
#include "mbedapi.h"

static const char *pinname[] = {
  "D0", "D1", "D2", "D3", "D4", "D5", "D6", "D7",
  "D8", "D9", "D10", "D11", "D12", "D13", "D14", "D15",
  "A0", "A1", "A2", "A3", "A4", "A5",
  "LED1", "LED2", "LED3", "LED4",
  "BUTTON1"
};
#define MBED_GPIO_PINS (sizeof(pinname)/sizeof(pinname[0]))

void
mrb_mruby_plato_gpio_grpeach_gem_init(mrb_state *mrb)
{
  size_t i;
  struct RClass *plato  = mrb_module_get(mrb, "Plato");
  struct RClass *gpio   = mrb_class_get_under(mrb, plato, "GPIO");

  for (i=0; i<MBED_GPIO_PINS; i++) {
    mrb_mod_cv_set(mrb, gpio, mrb_intern_cstr(mrb, pinname[i]), mrb_fixnum_value(i));
  }
}

void
mrb_mruby_plato_gpio_grpeach_gem_final(mrb_state *mrb)
{
}
